Adding ingester circuit breaker documentation (#8970)
grafana/mimir: docs/sources/mimir/configure/configure-ingester-circuit-breakers.md

---
aliases:
  - ../configuring/configuring-ingester-circuit-breakers/
description: Learn how to configure ingester circuit breakers.
menuTitle: Configure ingester circuit breakers
title: Configure Grafana Mimir ingester circuit breakers
---

# Configure Grafana Mimir ingester circuit breakers

## Background

> The circuit breaker pattern prevents an application from repeatedly trying to execute an operation that is likely to fail.

Reviewer: ```suggestion
Use circuit breakers to prevent an application from repeatedly trying to run an operation that is likely to fail.
```

Reviewer: We can make this more active and user-centric.

> A circuit breaker monitors the number of recent failures that have occurred, and uses this information to decide whether to allow a new operation to proceed, or simply return an exception immediately.

Reviewer: ```suggestion
A circuit breaker monitors the number of recent failures and then uses this information to decide whether to allow a new operation to proceed, or to simply return an exception immediately.
```

> In case of a failing operation, a circuit breaker allows an application to proceed with its execution, without waiting for the failure cause to be fixed.

Reviewer: ```suggestion
In the case of a failing operation, a circuit breaker allows an application to proceed, without waiting for you to resolve the failure cause.
```

> Since the failing operation is immediately rejected, the application doesn't retry to execute it, reducing this way its CPU usage.

Reviewer: ```suggestion
Because the failing operation is immediately rejected, the application doesn't retry it. This reduces its CPU usage.
```

> The circuit breaker pattern typically operates in 3 main states: `closed`, `open` and `half-open`.

Reviewer: ```suggestion
The circuit breaker pattern typically operates in 3 main states: `closed`, `open`, and `half-open`.
```

> - In the `closed` state, a circuit breaker operates normally, forwarding all the requests to the application it protects.

Reviewer: ```suggestion
- In the `closed` state, a circuit breaker operates normally, forwarding all requests to the application it protects.
```

- In the `open` state, the circuit breaker immediately stops forwarding requests to the failing application, effectively isolating it.

> - After a specified timeout period in the `open` state, the circuit breaker transitions to the `half-open`, where it forwards a limited number of trial requests to the application and monitors their execution.

Reviewer: ```suggestion
- After a specified timeout period in the `open` state, the circuit breaker transitions to the `half-open` state, where it forwards a limited number of trial requests to the application for monitoring.
```

Reviewer: This section adds a lot of great context! Thank you!

---
aliases:
  - ../configuring/configuring-ingester-circuit-breakers/
description: Learn how to configure ingester's circuit breakers.
menuTitle: Ingester's circuit breakers
title: Configure Grafana Mimir Ingester circuit breakers
---

# Configure Grafana Mimir Ingester circuit breakers

> ## How do Grafana Mimir Ingester circuit breakers work?

Reviewer 2: perhaps a section on why it would be a good idea to enable circuit breakers and when not to can help. That way people don't use it for things it wasn't designed and it (ideally) shows up as a search result when you google your Mimir problem

Author: Hi @dimitarvdimitrov, @tacole02 
I have added a background section, where I explained more details about the circuit breaker pattern.
Moreover, I explicitly stated what circuit breakers protect the ingesters from.
Could you please have a look and let me know what you think?

> A request to a resource protected by a circuit breaker follows these steps:

Author: I have fixed this.

### Read requests circuit breakers

Read requests circuit breakers follow the general procedure explained [above](#how-do-grafana-mimir-ingester-circuit-breakers-work), with the following remarks:

- in order to protect the write path as much as possible, ingesters do not allow read push requests if their push circuit breakers are open.
  This means that before an ingester tries to acquire a read circuit breaker permit, it first checks if its push circuit breaker is open.
  If it is the case, a circuit breaker open error is returned.
  Otherwise, the ingester tries to acquire a read requests circuit breaker permit.
- a read request that acquired a read requests circuit breaker permit meets the read requests circuit breaker failure condition if its duration is longer than the configurable read request maximum duration.

## Grafana Mimir Ingester circuit breakers configuration

> Grafana Mimir Ingester circuit breakers can be configured by using the following configuration options.

Reviewer 3: I don't think repeating the same sort of documentation that already exists for each config setting is useful. This also seems likely to go out-of-day. Maybe this could be a set of examples oriented around specific configurations you might want with a minimal set of flags required for that?

- `-ingester.push-circuit-breaker.enabled`: Enable circuit breaking when making push requests to ingesters.

- `-ingester.read-circuit-breaker.enabled`: Enable circuit breaking when making read requests to ingesters.

- `-ingester.push-circuit-breaker.thresholding-period`: Moving window of time that the percentage of failed requests is computed over for push circuit breakers.

- `-ingester.read-circuit-breaker.thresholding-period`: Moving window of time that the percentage of failed requests is computed over for read circuit breakers.

- `-ingester.push-circuit-breaker.failure-execution-threshold`: How many push requests must have been executed within the thresholding period for a push circuit breaker to be eligible to open for the rate of failures?

> - `-ingester.read-circuit-breaker.failure-execution-threshold`: How many read requests must have been executed within the thresholding period for a read circuit breaker to be eligible to open for the rate of failures?

Author: No, they shouldn't be questions.

- `-ingester.read-circuit-breaker.cooldown-period`: How long a read circuit breaker will stay in the open state before allowing additional read requests?

- `-ingester.push-circuit-breaker.initial-delay`: How long a push circuit breaker should wait between an activation request and becoming effectively active? During that time both failures and successes are not counted.

- `-ingester.read-circuit-breaker.initial-delay`: How long a read circuit breaker should wait between an activation request and becoming effectively active? During that time both failures and successes are not counted.

- `-ingester.push-circuit-breaker.request-timeout`: The maximum duration of a push request before it triggers a timeout. This configuration is used for push circuit breakers only, and the corresponding timeouts are not reported as errors.

- `-ingester.read-circuit-breaker.request-timeout`: The maximum duration of a read request before it triggers a timeout. This configuration is used for read circuit breakers only, and the corresponding timeouts are not reported as errors.

> ## Grafana Mimir Ingester circuit breakers metrics

Reviewer 3: Please make it clear that metrics aren't part of any API guarantee and can be changed at any time.

Grafana Mimir Ingester circuit breakers contain the following metrics:

> - `cortex_ingester_circuit_breaker_transitions_total`: a counter showing the number of times a circuit breaker has entered a state. It contains labels state (with possible values closed, open and half-open) and request_type (with possible values push and read).

Reviewer: ```suggestion
- `cortex_ingester_circuit_breaker_transitions_total`: Counter showing the number of times a circuit breaker enters a state. It contains the labels state, with possible values of `closed`, `open`, and `half-open`, and the `request_type` setting , with possible values of `push` and `read`.
```

> - `cortex_ingester_circuit_breaker_results_total`: a counter showing the results of executing requests via a circuit breaker. It contains lables result (with possble values success, error and circuit_breaker_open) and request_type (with possible values push and read).

Reviewer: ```suggestion
- `cortex_ingester_circuit_breaker_results_total`: Counter showing the results of executing requests via a circuit breaker. It contains lables result, with possible values of `success`, `error`, and `circuit_breaker_open`,  and the `request_type` setting, with possible values of `push` and `read`.
```

> - `cortex_ingester_circuit_breaker_request_timeouts_total`: a counter showing the number of times the circuit breaker recorded a request that reached timeout. It contains only the request_type label (with possible values push and read).

Reviewer: ```suggestion
- `cortex_ingester_circuit_breaker_request_timeouts_total`: Counter showing the number of times the circuit breaker records a request that reaches timeout. It contains the `request_type` label, with possible values of `push` and `read`.
```

---
aliases:
  - ../configuring/configuring-ingester-circuit-breakers/
description: Learn how to configure ingester circuit breakers.
menuTitle: Configure ingester circuit breakers
title: Configure Grafana Mimir ingester circuit breakers
---

# Configure Grafana Mimir ingester circuit breakers

## Background

Use circuit breakers to prevent an application from repeatedly trying to run an operation that is likely to fail.
A circuit breaker monitors the number of recent failures and then uses this information to decide whether to allow a new operation to proceed, or to simply return an exception immediately.
In the case of a failing operation, a circuit breaker allows an application to proceed, without waiting for you to resolve the failure cause.

> Because the failing operation is immediately rejected, the application doesn't retry it. This reduces the application's CPU usage.

Reviewer 2: can we say "Mimir" instead of "application?" Maybe even "the ingester" instead of "Mimir." It wasn't clear on a first read if an application is Mimir's client or Mimir itself.

I'd also mention somewhere (maybe not in this paragraph) that even if the operation fails in the ingester, the overall operation might still success because of the replication of data on ingesters. This is obvious to us, but may not be to someone new to Mimir.

Author: This is a general paragraph in which we explained how circuit breakers work. It is not related to Mimir only. That's why we opted for "application".

### Push requests circuit breakers

A push request meets the push requests circuit breaker failure condition if its duration is longer than the configured maximum push request duration.

### Read requests circuit breakers

Read requests circuit breakers follow these conditions:

- In order to protect the write path as much as possible, ingesters do not allow read push requests if their push circuit breakers are open.
  This means that before an ingester tries to acquire a read circuit breaker permit, it first checks if its push circuit breaker is open.
  If it's open, a circuit breaker open error is returned.
  Otherwise, the ingester tries to acquire a read requests circuit breaker permit.
- A read request that acquires a read requests circuit breaker permit meets the read requests circuit breaker failure condition if its duration is longer than the configured read request maximum duration.

## Grafana Mimir Ingester circuit breakers configuration

To enable Grafana Mimir ingester push and read circuit breakers, set `-ingester.push-circuit-breaker.enabled=true` and `-ingester.read-circuit-breaker.enabled=true`.

> You can enable the circuit breakers independently. Enabling one of them does not require enabling the other one.

Reviewer 2: any plans to make those on by default in some release? Maybe mention this here if yes

Author: I am not sure about this. For the moment circuit breakers are an experimental feature, and they will be disabled by default. In Grafana Cloud they are already enabled, and we are observing their behaviour. Depending on that we will decide how to proceed.
